File: easyPopup/src/main/java/org/fattili/easypopup/BasePopView.java
package org.fattili.easypopup;

import android.app.Activity;
import android.graphics.Rect;
import android.graphics.drawable.ColorDrawable;
import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.PopupWindow;
import android.widget.Toast;




/**
 * Created by dengzhenli on 2021/01/23.
 * 自定义的左侧浮窗
 */

public abstract class BasePopView {
    //修改的属性
    private int layoutId;
    private int popupWidth;
    private int popupHeight;

    private float marginWidth = 0;
    private float marginHeight = 0;


    //pop本地属性
    public BasePopupWindow popupWindow;
    public LayoutInflater inflate;
    public Activity activity;
    public View.OnClickListener onClickListener;
    private PopupDismissListener popupDismissListener;


    public BasePopView(Activity activity, View.OnClickListener clickListener) {
        this(activity);
        onClickListener = clickListener;
    }

    public BasePopView(Activity activity) {
        this.inflate = activity.getLayoutInflater();
        this.activity = activity;
    }


    /***********************************PopupWindow相关*********************************************/
    /**

    /**
     * 设置添加屏幕的背景透明度   * @param bgAlpha
     */
    private void backgroundAlpha(float bgAlpha) {
        WindowManager.LayoutParams lp = activity.getWindow().getAttributes();
        lp.alpha = bgAlpha; //0.0-1.0
        activity.getWindow().setAttributes(lp);
    }


    protected void initPopupWindowData(int popupWidth, int popupHeight, float ratiomarginWidth, float ratiomarginHeight) {
        this.popupHeight = popupHeight;
        this.popupWidth = popupWidth;
        this.marginHeight = ratiomarginHeight;
        this.marginWidth = ratiomarginWidth;
    }


    public boolean isShow() {
        if (popupWindow == null) return false;
        return popupWindow.isShowing();
    }


    /**
     * 初始化
     *
     * @param popupWidth
     * @param popupHeight
     * @param marginWidth
     * @param marginHeight
     */
    private void initPopupWindow(int popupWidth, int popupHeight, float marginWidth, float marginHeight) {
        initPopupWindowData(popupWidth, popupHeight, marginWidth, marginHeight);
        initPopupWindow();
    }

    /**
     * 初始化
     */
    private void initPopupWindow() {

        Log.d(TAG, "findResume: basepop：initPopupWindow");
        layoutId = getLayoutId();
        initPopupWindowData(popupWidth, popupHeight, marginWidth, marginHeight);
        final View popupWindowView = inflate.inflate(layoutId, null);
        //内容，高度，宽度
        popupWindow = new BasePopupWindow(popupWindowView, WindowManager.LayoutParams.WRAP_CONTENT, WindowManager.LayoutParams.FILL_PARENT, true);
        //动画效果
        popupWindow.setAnimationStyle(
                R.style.dialog_theme);

        //菜单背景色
        ColorDrawable dw = new ColorDrawable(0xff000000);

        View parent = activity.getWindow().getDecorView().findViewById(android.R.id.content);
        Rect rect = new Rect();
        Window window = activity.getWindow();
        window.getDecorView().getWindowVisibleDisplayFrame(rect);
        int height = rect.height();

//        popupWidth = activity.getResources().getDimensionPixelSize(LR.dimen.umipay_pop_width);
        popupWidth = 1000;
        popupHeight = height + rect.top;
        //宽度
        popupWindow.setWidth(popupWidth);
        //高度
        popupWindow.setHeight(popupHeight);
        //显示位置
        popupWindow.setOutsideTouchable(true);
        popupWindow.setBackgroundDrawable(dw);
        popupWindow.showAtLocation(parent, Gravity.TOP | Gravity.LEFT, (int) marginWidth, (int) marginHeight);

        //设置背景半透明
        backgroundAlpha(0.5f);


        popupWindow.setOnDismissListener(new PopupWindow.OnDismissListener() {
            @Override
            public void onDismiss() {
                onPopDismiss();
            }
        });

        initView(popupWindowView);
        initData();
    }

    private static final String TAG = "BasePopView";

    /**
     * 显示view
     */
    public void show() {
        Log.d(TAG, "findResume: basepop：show");

        if (popupWindow == null) {
            initPopupWindow();
        } else {
            View parent = activity.getWindow().getDecorView().findViewById(android.R.id.content);
            Rect rect = new Rect();
            Window window = activity.getWindow();
            window.getDecorView().getWindowVisibleDisplayFrame(rect);
            popupWindow.showAtLocation(parent, Gravity.TOP | Gravity.LEFT, (int) marginWidth, (int) marginHeight);
            reShowData();
        }
    }

    public interface PopupDismissListener {
        void onDismiss();
    }

    //隐藏
    public void dismiss() {
        if (popupWindow == null) return;
        popupWindow.dismiss();
    }

    public void onPopDismiss(){
        backgroundAlpha(1f);
        if (popupDismissListener != null) {
            popupDismissListener.onDismiss();
        }
    }

    public void finish() {
        if (popupWindow == null) return;
        if (popupWindow.isShowing()) {
            popupWindow.dismiss();
        }
        popupWindow = null;
    }

    public void showPW(BasePopView popView) {
        popView.show();
    }

    public void showMsg(String msg) {
    }

    //设置隐藏监听
    public void setPopupDismissListener(PopupDismissListener popupDismissListener) {
        this.popupDismissListener = popupDismissListener;
    }


    /**********************************抽象方法***********************************************************/
    public abstract void initView(View view);

    public abstract void initData();

    public void reShowData() {
        initData();
    }

    public abstract int getLayoutId();

}

File: easyPopup/src/main/java/org/fattili/easypopup/BasePopView.kt
package org.fattili.easypopup

import android.app.Activity
import android.content.Context
import android.graphics.Rect
import android.graphics.drawable.ColorDrawable
import android.util.Log
import android.view.*
import androidx.core.view.get

/**
 * Created by dengzhenli on 2021/01/23.
 * 自定义的左侧浮窗
 */
abstract class BasePopView(activity: Activity) {

    /**
     * layoutId
     */
    private var layoutId = 0

    /**
     * 窗体的长宽
     */
    private var popupWidth = WindowManager.LayoutParams.MATCH_PARENT
    private var popupHeight = WindowManager.LayoutParams.MATCH_PARENT

    /**
     * 窗体的长宽
     */
    private var viewWidth = WindowManager.LayoutParams.MATCH_PARENT
    private var viewHeight = WindowManager.LayoutParams.MATCH_PARENT

    /**
     * 上下边距
     */
    private var marginWidth = 0f
    private var marginHeight = 0f


    private var gravity = Gravity.TOP or Gravity.LEFT

    private var alpha = 0.5f

    /**
     * 是否占据焦点
     */
    private var focusable = true

    /**
     *
     */
    private var isOutsideTouchable = true


    //pop本地属性
    private val activity: Activity = activity
    private val context: Context = activity

    private var view: View? = null
    private var popupWindow: BasePopupWindow? = null

    /**********************************     抽象方法   **********************************/

    abstract fun onPopDismiss()

    abstract fun getLayoutId(): Int

    abstract fun initView(view: View?)

    abstract fun initData()

    private fun reShowData() {
        initData()
    }


    val isShow: Boolean
        get() = popupWindow?.isShowing ?: false

    companion object {
        private const val TAG = "BasePopView"
    }

    /***********************************PopupWindow相关 **********************************/

    private fun initPopData() {

    }

    /**
     * 初始化
     */
    private fun initPopupWindow() {
        layoutId = getLayoutId()
        val inflate = LayoutInflater.from(context)
        view = inflate.inflate(layoutId, null)
//        Log.d(TAG, "initPopupWindow: " + (view as ViewGroup).getChildAt(0).width)
        view?.let {


            val v = it.findViewWithTag<PopLayout>("PopLayout");
            viewWidth = it.measuredWidth
            viewHeight = it.measuredHeight
            popupWidth = it.measuredWidth
            popupHeight = it.measuredHeight
            Log.d(TAG, "initPopupWindow: " + v.width + ":" + it.width)
        }
        //内容，高度，宽度
        popupWindow = BasePopupWindow(
            view,
            viewWidth,
            viewHeight,
            focusable
        )

        //动画效果
        popupWindow?.animationStyle = R.style.dialog_theme
        //宽度
        popupWindow?.width = popupWidth
        //高度
        popupWindow?.height = popupHeight
        //显示位置
        popupWindow?.isOutsideTouchable = isOutsideTouchable
        popupWindow?.setBackgroundDrawable(view?.background)

        popupWindow?.showAtLocation(
            view,
            gravity,
            marginWidth.toInt(),
            marginHeight.toInt()
        )

        //设置背景半透明
        backgroundAlpha(alpha)
        popupWindow?.setOnDismissListener {
            backgroundAlpha(1f)
            onPopDismiss()
        }

        initView(view)
        initData()
    }


    /**
     * 显示view
     */
    fun show() {
        Log.d(TAG, "findResume: basepop：show")
        if (popupWindow == null) {
            initPopData()
            initPopupWindow()
        } else {
          popupWindow?.showAtLocation(
                view,
                gravity,
                marginWidth.toInt(),
                marginHeight.toInt()
            )
            reShowData()
        }
    }


    fun finish() {
        popupWindow?.let {
            if (it.isShowing) {
                it.dismiss()
            }
        }
        popupWindow = null
    }

    //隐藏
    fun dismiss() {
        popupWindow?.dismiss()
    }




    /**
     * 设置添加屏幕的背景透明度
     * @param bgAlpha
     */
    private fun backgroundAlpha(bgAlpha: Float) {
        activity.window.attributes.alpha = bgAlpha
    }
}
